RecoLocalTracker/SiStripClusterizer/plugins/SiStripRawToClusterGPUKernel.cu

```diff
@@ -0,0 +1,551 @@
+    const float minGoodCharge = sst_data_d->minGoodCharge;  //1620.0;
+    const auto nSeedStripsNC = clust_data_d->nClusters_;
+    const auto __restrict__ clusterIndexLeft = clust_data_d->clusterIndex_;
+
+    auto __restrict__ clusterSize = clust_data_d->clusterSize_;
+    auto __restrict__ clusterADCs = clust_data_d->clusterADCs_;
+    auto __restrict__ trueCluster = clust_data_d->trueCluster_;
+    auto __restrict__ barycenter = clust_data_d->barycenter_;
+    auto __restrict__ charge = clust_data_d->charge_;
+
+    constexpr uint16_t stripIndexMask = 0x7FFF;
+
+    const int tid = threadIdx.x;
+    const int bid = blockIdx.x;
+    const int nthreads = blockDim.x;
+    const int i = nthreads * bid + tid;
+
+    if (i < nSeedStripsNC) {
+      if (trueCluster[i]) {
+        const int left = clusterIndexLeft[i];
+        const int size = clusterSize[i];
+
+        if (i > 0 && clusterIndexLeft[i - 1] == left) {
+          trueCluster[i] = 0;  // ignore duplicates
+        } else {
+          float adcSum = 0.0f;
+          int sumx = 0;
+          int suma = 0;
+
+          auto j = 0;
+          for (int k = 0; k < size; k++) {
+            const auto index = left + k;
+            const auto chan = channels[index];
+            const auto fed = chanlocs->fedID(chan);
+            const auto channel = chanlocs->fedCh(chan);
+            const auto strip = stripId[index];
+#ifdef GPU_CHECK
+            if (fed == stripgpu::invalidFed) {
+              printf("Invalid fed index %d\n", index);
+            }
+#endif
+            if (strip != stripgpu::invalidStrip) {
+              const float gain_j = conditions->gain(fed, channel, strip);
+
+              uint8_t adc_j = adc[index];
+              const int charge = static_cast<int>(static_cast<float>(adc_j) / gain_j + 0.5f);
+
+              if (adc_j < 254) {
+                adc_j = (charge > 1022 ? 255 : (charge > 253 ? 254 : charge));
+              }
+              clusterADCs[j * nSeedStripsNC + i] = adc_j;
+
+              adcSum += static_cast<float>(adc_j);
+              sumx += j * adc_j;
+              suma += adc_j;
+              j++;
+            }
+          }  // loop over cluster strips
+          charge[i] = adcSum;
+          const auto chan = channels[left];
+          const fedId_t fed = chanlocs->fedID(chan);
+          const fedCh_t channel = chanlocs->fedCh(chan);
+          trueCluster[i] = (adcSum * conditions->invthick(fed, channel)) > minGoodCharge;
+          const auto bary_i = static_cast<float>(sumx) / static_cast<float>(suma);
+          barycenter[i] = static_cast<float>(stripId[left] & stripIndexMask) + bary_i + 0.5f;
+          clusterSize[i] = j;
+        }  // not a duplicate cluster
+      }    // trueCluster[i] is true
+    }      // i < nSeedStripsNC
+  }
+
+  void SiStripRawToClusterGPUKernel::unpackChannelsGPU(const ConditionsDeviceView *conditions, cudaStream_t stream) {
+    constexpr int nthreads = 128;
```

Comment:
hardcoded?

Reply:
Seems typical practice from a quick look through cmssw. What do you have in mind instead?

Reply:
@jpata ?

Reply:
cc @cms-sw/reconstruction-l2 :)

Reply:
yes - as I said, it seems to follow cmssw practice for this..

Reply:
We do plan a more general approach for configuring the number of threads and blocks; until it's OK to use a hard-coded value that gives a good performance.